docs/user_guide/source/components/anatomy.rst

```diff
@@ -0,0 +1,132 @@
+
+``Variable`` and ``Attribute`` definitions belong to one IO object, which means, they can only be used in one output.
+You need to define new ones for other outputs. Just because a Variable is defined, it will not appear in the output
+unless an associated Put() call provides the content.
+
+A stream is opened and closed once. The ``Engine`` object implements the data movement for the stream. It depends on the
+runtime options of the IO object that what type of an engine is created in the Open() call. One output step is denoted
+by a pair of BeginStep..EndStep block.
+
+An output step consist of variables and attributes. Variables are just definitions without content, so one must call a
+Put() function to provide the application data pointer that contains the data content one wants to write out. Attributes
+have their content in their definitions so there is no need for an extra call.
+
+Some rules:
+
+*   Variables can be defined any time, before the corresponding Put() call
+*   Attributes can be defined any time before EndStep
+*   The following functions must be treated as Collective operations
+
+  * ADIOS
+  * Open
+  * BeginStep
+  * EndStep
+  * Close
+
+.. note::
+
+    If there is only one output step, and we only want to write it to a file on disk, never stream it to other
+    application, then BeginStep and EndStep are not required but it does not make any difference if they are called.
+
+Anatomy of an ADIOS Input
+-------------------------
+
+.. code:: C++
+
+    ADIOS adios("config.xml", MPI_COMM_WORLD);
+    |
+    |   IO io = adios.DeclareIO(...);
+    |   |
+    |   |   Engine e = io.Open(...);
+    |   |   |
+    |   |   |   e.BeginStep()
+    |   |   |   |
+    |   |   |   |   varlist = io.AvailableVariables(...)
+    |   |   |   |   Variable var = io.InquireVariable(...)
+    |   |   |   |   Attribute attr = io.InquireAttribute(...)
+    |   |   |   |   |
+    |   |   |   |   |   e.Get(var, datapointer);
+    |   |   |   |   |
+    |   |   |   |
+    |   |   |   e.EndStep()
+    |   |   |
+    |   |   e.Close();
+    |   |
+    |   |--> IO goes out of scope
+    |
+    |--> ADIOS goes out of scope or adios2_finalize()
+
+The difference between input and output is that while we have to define the variables and attributes for an output, we
+have to retrieve the available variables in an input first as definitions (Variable and Attribute objects).
+
+If we know the particular variable (name and type) in the input stream, we can get the definition using
+InquireVariable(). Generic tools that process any input must use other functions to retrieve the list of variable names
+and their types first and then get the individual Variable objects. The same is true for Attributes.
+
+Anatomy of an ADIOS File-only Input
+-----------------------------------
+
+This only works for files on permanent storage for post-mortem processing. Do not use BeginStep/EndStep, so the
```

Comment:
This is not true anymore ever since BP5 became the default file engine, you need to open the file in RandomAccess mode to set the step selection.

Reply:
Yeah, I would introduce this section a bit differently.  Start with the two modes of ADIOS input, Read and ReadRandomAccess.  ReadRandomAccess can only be used with file engines and involves loading all the file metadata at once.  So it can be more memory intensive than Read mode, but allows reading data from any timestep using SetStepSelection().  Read mode allows data access only timestep by timestep, but generally it is more memory efficient as ADIOS is only required to load metadata for the current timestep.